src/main/java/org/terasology/web/io/JsonSession.java

```diff
@@ -101,63 +105,60 @@ public static void disconnectAllClients() {
+            }
+            instance.setResourceObservers();
+        });
     }
 
-    @SuppressWarnings("unchecked")
-    private void setResourceObservers() {
-        for (ObservableReadableResource observableResource: resourceManager.getAll(ObservableReadableResource.class)) {
-            observableResource.setObserver(client.getEntity(), resourceObserver);
-        }
-        for (EventEmittingResource eventResource: resourceManager.getAll(EventEmittingResource.class)) {
-            eventResource.setObserver(client.getEntity(), resourceObserver);
-        }
-        ResourceManager.getInstance().addEngineStateChangeObserver(engineStateObserver);
+    public void setResourceChangeSubscriber(BiConsumer<Collection<String>, JsonElement> resourceChangeSubscriber) {
+        this.resourceChangeSubscriber = resourceChangeSubscriber;
     }
 
-    private void removeResourceObservers() {
-        for (ObservableReadableResource observableResource: resourceManager.getAll(ObservableReadableResource.class)) {
-            observableResource.removeObserver(client.getEntity());
-        }
-        for (EventEmittingResource eventResource: resourceManager.getAll(EventEmittingResource.class)) {
-            eventResource.removeObserver(client.getEntity());
-        }
-        ResourceManager.getInstance().removeEngineStateChangeObserver(engineStateObserver);
+    public void setResourceEventListener(BiConsumer<Collection<String>, JsonElement> resourceEventListener) {
+        this.resourceEventListener = resourceEventListener;
     }
 
-    public void setReadableResourceObserver(BiConsumer<String, JsonElement> observer) {
-        resourceObserver.setReadableResourceObserver(observer);
+    private void notifyResourceChanged(ResourcePath resourcePath, Object newData) {
+        resourceChangeSubscriber.accept(Collections.unmodifiableCollection(resourcePath.getItemList()), GSON.toJsonTree(newData));
     }
 
-    public void setEventResourceObserver(BiConsumer<String, JsonElement> observer) {
-        resourceObserver.setEventResourceObserver(observer);
+    private void notifyResourceEvent(ResourcePath resourcePath, Object eventData) {
+        resourceEventListener.accept(Collections.unmodifiableCollection(resourcePath.getItemList()), GSON.toJsonTree(eventData));
+    }
+
+    private void setResourceObservers() {
+       resourceManager.addClient(client, this::notifyResourceChanged, this::notifyResourceEvent);
+    }
+
+    private void removeResourceObservers() {
+        resourceManager.removeClient(client);
     }
 
     public boolean isAuthenticated() {
         return client != null && !client.isAnonymous();
     }
 
     public ActionResult initAuthentication() {
-        /*if (isAuthenticated()) {
-            return new ActionResult(ActionResult.Status.UNAUTHORIZED, "Already authenticated");
-        }*/
+        if (isAuthenticated()) {
+            return new ActionResult(ActionResult.Status.FORBIDDEN, "Already authenticated");
```

Comment:
Yes I agree that a lot of the possible status codes make sense and I'm not really sure what's the best option... `ACTION_NOT_ALLOWED` maps to [405](https://httpstatuses.com/405) in the HTTP version which - in my understanding - would mean you can't POST /auth, however I don't know - is it meant only for situations where the server *never* supports that method for the requested resource, or it's OK also if it doesn't support it *now*? I suspect there isn't a correct answer because the authentication isn't really stateless (the state is the fact that you are anonymous or authenticated). Also, there is nothing an authenticated client can do with `/auth` (which supports GET and POST for unauthenticated clients, respectively to retrieve the server handshake hello message and to perform the authentication). For `OK` I'm not sure, because, I think (but please correct me if I'm wrong) in the GET part (i.e. an authenticated client trying to redo the authentication handshake) the server should still send a response content.